Disable CRC32-VX Extention for some Clang versions

fneddy opened the pull request:
We have to disable the CRC32-VX implementation for some Clang versions (18 <= version < 19.1.2) that generate bad code for the IBM S390 VGFMA intrinsics.

This fixes #1845

Question: should we print a warning during configure, if the intrinsics are not used. Or is it OK to just silently disable them?

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Bug Fixes**
	- Added compiler version compatibility checks for CRC32-VX functionality on S390 architecture.
	- Improved runtime CPU detection and CRC32 function selection for specific compiler versions. 
	- Provided user guidance for incompatible Clang versions regarding CRC32-VX extension usage.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Dead2 commented:
@fneddy Codefactor had a couple suggestions to improve portability and performance of the changes to Configure, would be good if you considered these changes, both suggestions might potentially otherwise become parts of a future fix/cleanup PR if someone decides to do some janitorial work on Configure.

Other than that, this looks pretty good to me. Have not tested though.

nmoinvaz commented on CMakeLists.txt:
```diff
@@ -875,7 +875,11 @@ if(WITH_OPTIM)
             check_vgfma_intrinsics()
-            if(HAVE_VGFMA_INTRIN)
+            if(HAVE_VGFMA_INTRIN AND
+                NOT ("${CMAKE_C_COMPILER_ID}" STREQUAL "Clang" AND
```
No need for quotes or ${] around CMAKE_C variables. See other examples in this cmake file.

nmoinvaz commented on configure:
```diff
@@ -1499,6 +1504,20 @@ EOF
     fi
+
+    printf "Checking compiler version for good s390 vector extension... " | tee -a configure.log
+    CC_STRING=$( $CC --version  | grep -oE 'clang version [0-9]+\.[0-9]+(\.[0-9]+)?')
```
Extra spaces after $( on next two lines. and Extra space before ending ).

phprus commented on CMakeLists.txt:
```diff
@@ -875,7 +875,11 @@ if(WITH_OPTIM)
-            if(HAVE_VGFMA_INTRIN)
+            if(HAVE_VGFMA_INTRIN AND
+                NOT ("${CMAKE_C_COMPILER_ID}" STREQUAL "Clang" AND
+                "${CMAKE_C_COMPILER_VERSION}" VERSION_GREATER_EQUAL 18 AND
```
``VERSION_GREATER_EQUAL`` - cmake 3.7+, but zlib-ng required cmake 3.5+.

fneddy replied:
i guess bumping the version is not an option. so i'll split it into two checks.

mtl1979 commented on configure:
```diff
@@ -1465,6 +1465,11 @@ EOF
 }
 
+version_number() {
+    # convert a version from string into a numberical, assuming 8 bits per nibble
```
"a number", or "numerical", not "a numberical".

fneddy commented on configure:
```diff
@@ -1499,6 +1504,20 @@
+    CC_ID=$(echo ${CC_STRING} | awk '{print $1}')
+    CC_VERSION=$(version_number $( echo ${CC_STRING} | awk '{print $3}'))
+
+    if ([ "${CC_ID}" = "clang" ] &&
+        [ ${CC_VERSION} -ge $(version_number 18.0.0) ] &&
+        [ ${CC_VERSION} -lt $(version_number 19.1.2) ]); then
+        HAVE_VGFMA_INTRIN=0
+        echo "No." | tee -a configure.log
+    else
+        echo "Yes." | tee -a configure.log
+    fi
```
What do the **non AIs** think about this? Personally I  would prefer to not spam the log with this much output.

Dead2 commented:
@fneddy Regarding warning the user about the problem with the clang version, I do prefer to not hide it.
Ref my comment on #1845
`Either that or refuse to compile at all with VX enabled on those versions, so it is not a silent degradation.`

Personally I would actually prefer it be a visible warning or even an error to enabling VX on known broken Clang versions.
It can be useful information for the person compiling it, and it does not affect the runtime library. It should however not log a message on the good Clang versions IMO, unlike what typically is done with autotools configure.

Configure and CMake should have the same behavior for this.

**Suggestion for discussion:** If error is the way to go, then this check could perhaps instead be done in `s390_functions.h` to error out of compilation. The only upside of that is that it could be done in one place and it would also work outside of CMake and Configure (not that there is likely to be any others, NMake is probably not used for s390x I assume).

fneddy committed:
Disable CRC32-VX Extention for some Clang versions
We have to disable the CRC32-VX implementation for some Clang versions
(18 <= version < 19.1.2) that generate bad code for the IBM S390 VGFMA intrinsics.

nmoinvaz commented on arch/s390/s390_functions.h:
```diff
@@ -5,10 +5,22 @@
+
+#ifdef __clang__
+#  if ( (__clang_major__ >=18) && (__clang_major__ <= 19 && __clang_minor__ <= 1 && __clang_patchlevel__ < 2) )
+#    pragma message "CRC32-VX extension is broken on current Clang version (" XSTR(__clang_major__) "." XSTR(__clang_minor__) "." XSTR(__clang_patchlevel__) "). " \
```
I wouldn't bother trying to print the current version. It would remove the need for the macros.

fneddy commented on arch/s390/s390_functions.h:
```diff
@@ -7,8 +7,16 @@
 #ifdef S390_CRC32_VX
 uint32_t crc32_s390_vx(uint32_t crc, const uint8_t *buf, size_t len);
+
+#ifdef __clang__
+#  if ( (__clang_major__ >=18) && (__clang_major__ <= 19 && __clang_minor__ <= 1 && __clang_patchlevel__ < 2) )
+#   error "CRC32-VX extension is broken on current Clang version. " \
+          "Affected Clang versions are: 18.0.0 <= clang_version < 19.1.2. " \
+          "Either disable CRC32-VX extension or switch to another compiler."
+#  endif
```
please explain what versions are not catched by my check.

phprus commented on arch/s390/s390_functions.h:
```diff
@@ -7,8 +7,16 @@
 uint32_t crc32_s390_vx(uint32_t crc, const uint8_t *buf, size_t len);
+
+#ifdef __clang__
+#  if ( (__clang_major__ == 18) || (__clang_major__ == 19 && __clang_minor__ <= 1 && __clang_patchlevel__ < 2) )
```
```c
#if ((__clang_major__ == 18) || (__clang_major__ == 19 && (__clang_minor__ < 1 || (__clang_minor__ == 1 && __clang_patchlevel__ < 2))))
```

fneddy replied:
ufff ... yes you are right

Dead2 commented:
@fneddy If you could do a rebase, that'd be great. CI is too broken to see anything useful after github actions image update.

fneddy commented:
ok, hold on. I fix this.

Dead2 commented:
This tested the code nicely.
https://github.com/zlib-ng/zlib-ng/actions/runs/12925714333/job/36047299493?pr=1852

Seems like quotation is not needed for the preprocessor error directive, my suggestion for a perhaps clearer error:
```
#   error CRC32-VX optimizations are broken due to compiler bug in Clang versions: 18.0.0 <= clang_version < 19.1.2. \
          Either disable the zlib-ng CRC32-VX optimization, or switch to another compiler/compiler version.
```
What do you think?

fneddy commented:
yes that looks indeed better! thx!

Dead2 reviewed (approved):
LGTM
